test/test_parameter_defs_csv.py

```diff
@@ -203,6 +211,81 @@ def test_parameter_function_map(self):
 
         self.assertEqual(len(errors), 0, msg=errors)
 
+    def test_optional_args(self):
+        """
+        Parameter Function Map Optional Args - If the string 'None' is specified
+        as a possible argument to a function, Stream Engine will consider that
+        argument to be optional and will call the function even if a Parameter
+        could not be found for that argument. However, the actual function that is
+        called should specify default value for that arg or the interpreter will
+        not be able to resolve the function signature during the call.
+        """
+        errors = []
+
+        # Start with rows with both the parameter function id and parameter
+        # function map are populated.
+        idx = ((self.data.parameterfunctionid != '') & (self.data.parameterfunctionmap != ''))
```

Comment:
This line is concerning because:

* the code looks and smells like C
* if neither value is set, it will return 0 - is that the desired result?
* it's difficult to see if this is creating a single value or doing a clever tuple generator expression later on you are looping over data. Is this an operation on a numpy array?

Maybe I'm just not hip on numpy array processing, but is seems like the line could be more like:

`idx = self.data.parameterfunctionid if self.data.parameterfunctionid else self.data.parameterfunctionmap`

And, if so, that doesn't match the comment or the for loop below that seems to loop of the matches.

Reply:
idx is a numpy boolean index array. This boolean array is used as a mask to filter the data array in the next step. I have fixed the comments below.